base/abstractset.jl

```diff
@@ -119,7 +119,15 @@ Set{Int64} with 1 element:
 """
 intersect(s::AbstractSet, itr, itrs...) = intersect!(intersect(s, itr), itrs...)
 intersect(s) = union(s)
-intersect(s::AbstractSet, itr) = mapfilter(_in(s), push!, itr, emptymutable(s))
+function intersect(s::AbstractSet, itr) 
+    if haslength(itr) && hasfastin(itr)
+        if length(s) > length(itr) 
+            return mapfilter(_in(s), push!, itr, emptymutable(s)) 
+        else
+            return mapfilter(_in(itr), push!, s, emptymutable(s))
+        end
+    end
```

Comment:
You need something like

```suggestion
    if hasfastin(itr) && haslength(itr) && length(s) < length(itr) 
        return mapfilter(_in(itr), push!, s, emptymutable(s))
    else
        return mapfilter(_in(s), push!, itr, emptymutable(s)) 
    end
```

Reply:
Also please make sure both branches are tested.

Reply:
sorry for the delay, it should be fine now also added tests for it